src/rng/src/main.rs

```diff
@@ -0,0 +1,192 @@
+
+use clap::{load_yaml, App};
+use std::fs::File;
+use std::path::Path;
+use std::sync::{Arc, Mutex, RwLock};
+use std::thread;
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vhu_rng::VuRngBackend;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+fn start_backend(cmd_args: clap::ArgMatches, dry_run: bool) -> Result<(), String> {
+    let mut handles = Vec::new();
+
+    let path = cmd_args
+        .value_of("socket_path")
+        .ok_or("Invalid socket path")?;
+
+    let count = cmd_args
+        .value_of("socket_count")
+        .unwrap_or("1")
+        .parse::<u32>()
+        .map_err(|_| "Invalid socket_count")?;
+
+    let source = match cmd_args.value_of("rng_source") {
+        Some(source) => {
+            if Path::new(source).is_file() {
+                source.to_string()
+            } else {
+                return Err(String::from("Enable to access RNG source file"));
+            }
+        }
+        None => "/dev/urandom".to_string(),
+    };
+
+    let random_file = match File::open(source) {
+        Ok(random_file) => Arc::new(Mutex::new(random_file)),
+        Err(e) => return Err(e.to_string()),
+    };
+
+    let period_ms: u128 = match cmd_args.value_of("period") {
+        Some(period_ms) => match period_ms.parse::<u128>() {
+            Ok(value) => {
+                if value > (1 << 16) {
+                    return Err(format!(
+                        "Input period too big, maximum value is {}",
+                        1 << 16
+                    ));
+                }
+                value
+            }
+            Err(e) => {
+                return Err(e.to_string());
+            }
+        },
+        None => 1 << 16,
+    };
+
+    let mut max_bytes: usize = match cmd_args.value_of("max_bytes") {
+        Some(max_bytes) => match max_bytes.parse::<usize>() {
+            // No point in checking for a maximum value like above,
+            // the library parsing code will do that for us.
+            Ok(value) => value,
+            Err(_) => {
+                return Err(String::from("Enable to process max-bytes"));
+            }
+        },
+        None => usize::MAX,
+    };
+
+    // Divide available bandwidth by the number of threads in order
+    // to avoid overwhelming the HW.
+    max_bytes /= count as usize;
+
+    for i in 0..count {
+        let socket = path.to_owned() + &i.to_string();
+        let random = Arc::clone(&random_file);
+
+        let handle = thread::spawn(move || loop {
+            let rng_backend = match VuRngBackend::new(random.clone(), period_ms, max_bytes) {
+                Ok(rng_backend) => rng_backend,
+                Err(_) => {
+                    println!("Error creating RNG backend on thread: {}", i);
+                    return;
+                }
+            };
+
+            let vu_rng_backend = Arc::new(RwLock::new(rng_backend));
+
+            if dry_run {
+                return;
+            }
```

Comment:
What is the purpose of `dry_run`? Do we want to use it in unit tests? Depending on what we want to achieve with it, we should look at other ways of achieving the same behavior.

What is important to understand here is what functionality is the one we need to create a mock for, and then make sure that the interface of the function receives that functionality as a parameter. This is only possible once we break down this function in smaller functions with clear responsibilities.

Reply:
@mathieupoirier look at my latest pull request. I have fixed this stuff in there and made other general improvements.